ckan/logic/action/file.py

```diff
@@ -0,0 +1,775 @@
+    """Add user from context as file owner."""
+    cache = logic.ContextCache(context)
+    user = cache.get("user", context["user"], lambda: model.User.get(context["user"]))
+    if user:
+        owner = model.Owner(
+            item_id=item_id,
+            item_type=item_type,
+            owner_id=user.id,
+            owner_type="user",
+        )
+        context["session"].add(owner)
+
+
+def _process_filters(  # noqa: C901
+    filters: dict[str, Any], columns: Mapping[str, Column[Any]]
+) -> ColumnElement[bool]:
+    """Transform `{"$and":[{"field":{"$eq":"value"}}]}` filters into SQL filters."""
+
+    items = []
+
+    for k, v in filters.items():
+        if k in ["$and", "$or"]:
+            if not isinstance(v, list):
+                raise logic.ValidationError(
+                    {"filters": [f"Only lists are allowed inside {k}"]}
+                )
+            nested_items = [
+                _process_filters(sub_filters, columns)
+                for sub_filters in v
+                if isinstance(sub_filters, dict)
+            ]
+            if len(nested_items) > 1:
+                wrapper = sa.and_ if k == "$and" else sa.or_
+                items.append(wrapper(*nested_items).self_group())
+            else:
+                items.extend(nested_items)
+
+        elif k in ["storage_data", "plugin_data"]:
+            items.extend(_process_data(columns[k], v))
+
+        elif k in columns:
+            items.extend(_process_field(columns[k], v))
+
+        else:
+            raise logic.ValidationError({"filters": [f"Unknown filter: {k}"]})
+
+    if len(items) == 1:
+        return items[0]
+
+    return sa.and_(*items).self_group()
+
+
+_op_map = {
+    "$eq": "=",
+    "$ne": "!=",
+    "$lt": "<",
+    "$lte": "<=",
+    "$gt": ">",
+    "$gte": ">=",
+    "$in": "IN",
+    "$is": "IS",
+    "$isnot": "IS NOT",
+    "$like": "LIKE",
+    "$ilike": "ILIKE",
+}
+
+
+def _process_field(col: Column[Any], value: Any):  # noqa: C901
+    """Transform `{"field":{"$eq":"value"}}` into SQL filters."""
+    if isinstance(value, list):
+        value = {"$in": value}
+
+    elif value is None:
+        value = {"$eq": None}
+
+    elif not isinstance(value, dict):
+        value = {"$eq": value}
+
+    for operator, filter in value.items():
+        column = col
+        if operator not in _op_map:
+            raise logic.ValidationError(
+                {"filters": [f"Operator {operator} is not supported"]}
+            )
+        op = _op_map[operator]
+        if filter is None:
+            if op == "=":
+                op = "is"
+            elif op == "!=":
+                op = "IS NOT"
+
+        elif operator == "$in" and isinstance(filter, list):
+            filter = tuple(filter)
+        elif not isinstance(filter, col.type.python_type):
+            filter = str(filter)
+            column = sa.func.cast(column, sa.Text)
+
+        func = column.bool_op(op)
+        yield func(filter)
```

Comment:
when https://github.com/ckan/ckan/pull/9028 is merged we can share same filters.py implementation from @amercader